core/src/main/java/com/linecorp/armeria/server/HttpHeaderInjectingRpcService.java

```diff
@@ -0,0 +1,129 @@
+ * A Thrift call replies with an {@link RpcResponse} which is wrapped by an {@link HttpResponse}.
+ * The {@link HttpHeaders} returned from {@link #httpHeaders(ServiceRequestContext, RpcRequest, RpcResponse)}
+ * will be added to that {@link HttpResponse}.
+ *
+ * <pre>{@code
+ * final ServerBuilder sb = ...
+ * sb.service("/", ThriftCallService.of(thriftHandler)
+ *                                  .decorate(...)
+ *                                   // A subclass of HttpHeaderInjectingRpcService
+ *                                  .decorate(MyInjectingService::new)
+ *                                  .decorate(THttpService.newDecorator()));
+ * }</pre>
+ *
+ * <p>Note that this class adds {@link HttpHeaders} only when the {@link HttpRequest} is decoded to
+ * an {@link RpcRequest} successfully. If you want to add {@link HttpHeaders} regardless of the decoding
+ * succeeded or not, use your own decorator:
+ *
+ * <pre>{@code
+ * > public final class MyOwnInjector extends SimpleDecoratingService<HttpRequest, HttpResponse> {
+ * >     MyOwnInjector(Service<HttpRequest, HttpResponse> delegate) {
+ * >         super(delegate);
+ * >     }
+ * >
+ * >     @Override
+ * >     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+ * >         final HttpResponse originalRes = delegate().serve(ctx, req);
+ * >         final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+ * >         final HttpResponse newRes = HttpResponse.from(future);
+ * >
+ * >         originalRes.aggregate().whenComplete((res, cause) -> {
+ * >             if (cause != null) {
+ * >                 future.completeExceptionally(cause);
+ * >             } else {
+ * >                 final HttpHeaders headers =
+ * >                 res.headers().add(AsciiString.of("my-header"), "foo");
+ * >                 future.complete(HttpResponse.of(headers, res.content(), res.trailingHeaders()));
+ * >             }
+ * >         });
+ * >         return newRes;
+ * >     }
+ * > }
+ * }</pre>
+ *
+ * <p>Then, decorate it:
+ *
+ * <pre>{@code
+ * final ServerBuilder sb = ...
+ * sb.service("/", ThriftCallService.of(thriftHandler)
+ *                                  .decorate(...)
+ *                                  .decorate(THttpService.newDecorator())
+ *                                  .decorate(MyOwnInjector::new));
+ * }</pre>
+ */
+public abstract class HttpHeaderInjectingRpcService extends SimpleDecoratingService<RpcRequest, RpcResponse> {
+
+    /**
+     * Creates a new instance that decorates the specified {@link Service}.
+     */
+    protected HttpHeaderInjectingRpcService(Service<RpcRequest, RpcResponse> delegate) {
+        super(delegate);
+    }
+
+    @Override
+    public RpcResponse serve(ServiceRequestContext ctx, RpcRequest req) throws Exception {
+        final RpcResponse originalRes = delegate().serve(ctx, req);
```

Comment:
How about using `RpcResponse#from` like below? Because there is some duplicated code with that. Then, we may change `httpHeaders` method's signature to `httpHeaders(ServiceRequestContext ctx, RpcRequest req, @Nullable Object res, @Nullable Throwable cause)`.

```
        return RpcResponse.from(delegate().serve(ctx, req).handle((res, cause) -> {
            final HttpHeaders headers = httpHeaders(ctx, req, res, cause);
            if (headers != null && !headers.isEmpty()) {
                ctx.addAdditionalResponseHeaders(headers);
            }
            return res;
        }));
```

Reply:
I duplicate the codes because there were some issued regarding the execution order of `CompletableFuture`s callbacks. But, because I changed the logic to inject `HttpHeaders` in `HttpResponseSubscriber`, this would be much nicer. Thank you~!